Plumb futures through caching/spawning layer, leverage error-chain

alexcrichton opened the pull request:
This PR builds on https://github.com/mozilla/sccache/pull/68 by plumbing futures all the way through the caching layer, effectively future-izing the entire server. The master branch of hyper is used for HTTP requests and tokio-process is used for process management.

This also cleans up a bit along the way related to error handling by leveraging error-chain in as many places as possible. To facilitate that an `SFuture` ("sccache future") alias was introduced to easily pass around and return types of futures.

Note that this should probably hold off on merging until https://github.com/mozilla/sccache/pull/68 lands, but I figured it'd be interesting at least!

luser commented:
I had started on some work in the error-chain vein recently: https://github.com/luser/sccache/tree/error-chain

I held off landing it because I figured it was going to conflict with the tokio rewrite. The only functional change there is that I changed `notify_server_startup` to send an error message over the socket/pipe if startup fails.

alexcrichton commented:
Oh yeah no worries! All those changes are just in the last commit, so I don't mind backing that out as well if you'd prefer

luser commented:
The Rust 1.13 Travis builds all seem to be [failing with errors like](https://travis-ci.org/mozilla/sccache/jobs/201903329#L302) `error[E0252]: a type named ErrorKind has already been imported in this module`.

The Windows builds in AppVeyor [are failing with](https://ci.appveyor.com/project/luser/sccache2/build/1.0.177/job/aijrsrtx8k0ehcpp#L368):
```
error[E0277]: the trait bound `*mut mio::poll::ReadinessNode: std::marker::Sync` is not satisfied
   --> src\mock_command.rs:147:6
```

On other builds there are some test failures. It looks like `test_server_port_in_use` is failing on all builds and `test_compiler_get_cached_or_compile_cached` failed in at least one instance.

alexcrichton commented:
Oh dear, how embarassing! Looks like I was working with a stale sccache binary locally with up to date tests by accident. I think I've fixed everything up, but I'll check in to make sure

alexcrichton commented:
Ok, that last commit should do the trick, I can rebase the changes to the right places at the end.

luser commented on src/compiler/compiler.rs:
```diff
@@ -275,124 +296,205 @@ impl Compiler {
+                for (key, path) in &outputs {
+                    let mut f = try!(File::open(&path));
+                    try!(entry.put_object(key, &mut f).map_err(|e| {
+                        let msg = format!("failed to put object `{:?}` in \
```
You lost the `out_file` from the error message here, which is not a huge problem but I put those in all the log messages because otherwise trying to interpret the sccache log from a real parallel build was a nightmare.

alexcrichton replied:
Oh good point yeah. I've been trying to leverage error-chain now to ensure that all the errors returned have all necessary context on them, but I'll be sure to leave this in as well.

luser commented on src/compiler/compiler.rs:
```diff
@@ -275,124 +296,205 @@ impl Compiler {
+        let start = Instant::now();
+
+        let compile = self.kind.compile(creator, self, stdout, &parsed_args, cwd, &pool);
+        Box::new(compile.and_then(move |(cacheable, compiler_result)| {
```
Just curious: is there a reason to prefer intermediate variables like `compile` here and `write` below vs. just chaining `and_then` off the previous method?

alexcrichton replied:
Ah this is just stylistic for me. I prefer to avoid rightward-drift and keep things under 80 cols, so I tend to make more local variables to avoid the drift and indentation. Both work equivalently though of course!

luser commented on src/compiler/compiler.rs:
```diff
@@ -275,124 +296,205 @@ impl Compiler {
+                }
+
+                // Try to finish storing the newly-written cache
+                // entry. We'll get the result back elsewhere.
```
This comment should probably be clearer that we don't want to block returning the compile result to the client while we write to the cache.

alexcrichton commented on src/compiler/compiler.rs:
```diff
@@ -275,124 +296,205 @@ impl Compiler {
+                    stdout: vec!(),
+                    ..preprocessor_result
+                };
+                return Box::new(future::ok((CompileResult::Error, output))) as Box<Future<Item=_, Error=_>>
```
Most of the time, yeah, but `.boxed()` has the added requirement of `Send` so here where most trait objects are not `Send` we can't always use it. In this case it inferred the closure return type as then a future which was `Send` causing errors for later returns.

There's other annotations that can happen here, though, so I'll apply those.

alexcrichton commented on src/compiler/compiler.rs:
```diff
@@ -275,124 +296,205 @@ impl Compiler {
+                            stderr: stderr.into_inner(),
+                        };
+                        let result = CompileResult::CacheHit(duration);
+                        return Box::new(write.map(|_| {
```
Heh yeah it took me awhile to get the hang of them but I've found them to be super useful  once you get the hang

alexcrichton commented on src/compiler/compiler.rs:
```diff
@@ -275,124 +296,205 @@ impl Compiler {
+                    Err(e) => {
+                        debug!("[{}]: Cache read error: {:?}", parsed_args.output_file(), e);
                         //TODO: store the error in CacheReadError in some way
                         MissType::CacheReadError
```
Sure yeah, I'll implement that

alexcrichton commented:
Ok, updated!

luser commented on src/mock_command.rs:
```diff
@@ -142,84 +149,118 @@ impl CommandChild for Child {
     }
 }
 
 /// Unit struct to use `RunCommand` with `std::system::Command`.
```
This doc is no longer correct.

alexcrichton commented on src/commands.rs:
```diff
@@ -471,65 +474,62 @@ fn handle_compile_finished<T : Write, U : Write>(response : CompileFinished, std
+            }
+            Ok(output.status.code()
+               .unwrap_or_else(|| {
+                   /* TODO: this breaks type inference, figure out why
```
Heh, will do!

alexcrichton commented on src/compiler/clang.rs:
```diff
@@ -23,6 +23,7 @@ use ::compiler::{
     run_input_output,
 };
 use futures::Future;
+use futures::future;
```
Oh sure, I'll change to `use futures::future::{self, Future}`

alexcrichton commented on src/compiler/clang.rs:
```diff
@@ -57,49 +58,69 @@ pub fn compile<T>(creator: &T,
+    // Clang needs a temporary file for compilation, otherwise debug info
+    // doesn't have a reference to the input file.
+    let input = parsed_args.input.clone();
+    let write = pool.spawn_fn(move || {
```
Excellent idea! I count 4 locations which had to independently remember "oh don't forget to *not* drop `TempDir` early"!

alexcrichton commented on src/compiler/clang.rs:
```diff
@@ -57,49 +58,69 @@ pub fn compile<T>(creator: &T,
+    let cwd2 = cwd.to_string();
+    let mut creator2 = creator.clone();
+    let parsed_args2 = parsed_args.clone();
+    let out_file2 = out_file.clone();
```
Hm let me see if I can clean this up, I've actually got an idea to make this better.

luser commented on src/simples3/credential.rs:
```diff
@@ -309,180 +314,178 @@ fn parse_credentials_file(file_path: &Path) -> Result<HashMap<String, AwsCredent
+        let mut req = Request::new(Method::Get, address.parse().unwrap());
+        req.headers_mut().set(Connection::close());
+        let response = self.client.request(req).and_then(|response| {
+            response.body().fold(Vec::new(), |mut body, chunk| {
```
Does hyper not have a helper method for "collect the response body into a Vec of bytes"?

alexcrichton replied:
I don't believe so, but https://github.com/hyperium/hyper/issues/953 is related

luser replied:
Actually, given that `hyper::Body` is just implementing `Stream`, maybe there should be an extension trait in futures to help with this, something like:
```rust
trait ByteStreamExt {
  type Error;
  fn collect_bytes(self) -> Future<Item = Vec<u8>, Error = Self::Error>;
}

impl<T, I, E> ByteStreamExt for T where T: Stream<Item = I, Error = E, I: AsRef<u8> {
// ...
}
```

alexcrichton replied:
That actually sounds pretty nifty to me! I've filed https://github.com/alexcrichton/futures-rs/issues/390

alexcrichton commented:
Ok I was able to cut down on most of the `let foo2 = foo.clone()`, turns out nothing goes wrong if the `Command` is created *outside* the `and_then` callback! That's just a few internal allocations to hold the description in memory, but the actual spawning is deferred until later.

luser commented on src/simples3/credential.rs:
```diff
@@ -309,180 +314,178 @@ fn parse_credentials_file(file_path: &Path) -> Result<HashMap<String, AwsCredent
+        let body = body.map_err(|_e| {
+            CredentialsError::new("Didn't get a parseable response body from instance role details")
+        }).and_then(|body| {
+            String::from_utf8(body).map_err(|_| {
```
These are a little clunky, but I haven't looked at the error-chain patch yet.

luser commented on src/cache/disk.rs:
```diff
@@ -57,41 +57,39 @@ fn make_key_path(key: &str) -> PathBuf {
-                 })
+            let v = entry.finish()?;
+            lru.lock().unwrap().insert_bytes(key, &v)?;
+            Ok(start.elapsed())
```
That is significantly nicer!

luser commented on src/commands.rs:
```diff
@@ -100,94 +100,92 @@ fn get_port() -> u16 {
-                            return Ok(ServerStartup::TimedOut);
-                        }
-                    }
+    let listener = UnixListener::bind(&socket_path)?;
```
I guess we could replace most of this with tokio-uds, right? (It'd be nice to get rid of some unsafe code.)

alexcrichton replied:
Certainly!

luser commented on src/commands.rs:
```diff
@@ -206,11 +204,13 @@ fn redirect_stderr(f: File) -> io::Result<()> {
+        _ => return Ok(()),
+    };
+    let f = OpenOptions::new().create(true).append(true).open(name)?;
+    redirect_stderr(f)
```
We apparently have inherently different styles about using intermediate variables. :) I think I err on the "way too long functional programming" side of things. Probably all that LISP programming in my formative years!

alexcrichton replied:
Oh I certainly don't mind changing back, I think this was primarily just line-length limits for myself which I tend to be a little more conservative on than most.

luser replied:
Nah, it's fine, my code gets way *too* long.

luser commented on src/commands.rs:
```diff
@@ -235,148 +235,137 @@ fn run_server_process() -> io::Result<ServerStartup> {
+    };
+    let mut si: STARTUPINFOW = unsafe { mem::zeroed() };
+    si.cb = mem::size_of::<STARTUPINFOW>() as DWORD;
+    if unsafe { kernel32::CreateProcessW(exe.as_mut_ptr(),
```
I'm looking forward to removing all this code when 1.16 releases...

luser commented on src/commands.rs:
```diff
@@ -235,148 +235,137 @@ fn run_server_process() -> io::Result<ServerStartup> {
         Ok(response.take_stats())
     } else {
-        Err(Error::new(ErrorKind::Other, "Unexpected server response!"))
+        Err("Unexpected server response!".into())
```
This can just be `bail!("...")` (same thing elsewhere)

luser commented on src/errors.rs:
```diff
@@ -12,8 +12,44 @@
+pub type SFuture<T> = Box<Future<Item = T, Error = Error>>;
+
+pub trait FutureChainErr<T> {
+    fn chain_err<F, E>(self, callback: F) -> SFuture<T>
```
This is nice, chain-err should probably just have this built-in.

alexcrichton replied:
Certainly, yeah, I opened https://github.com/brson/error-chain/issues/130

luser commented on src/cache/cache.rs:
```diff
@@ -87,15 +85,19 @@ pub struct CacheRead {
 impl CacheRead {
     /// Create a cache entry from `reader`.
     pub fn from<R: ReadSeek + 'static>(reader: R) -> io::Result<CacheRead> {
```
We could just make this return `error::Result<CacheRead>` and add `zip::result::ZipResult` to the error-chain.

luser commented on src/cache/s3.rs:
```diff
@@ -51,7 +46,9 @@ pub struct S3Cache {
 impl S3Cache {
     /// Create a new `S3Cache` storing data in `bucket`.
     pub fn new(bucket: &str, endpoint: &str, handle: &Handle) -> io::Result<S3Cache> {
```
Similarly this could be using error-chain.

luser commented on src/compiler/compiler.rs:
```diff
@@ -424,39 +417,33 @@ impl Compiler {
                        parsed_args.output_file());
                 return Box::new(future::ok((CompileResult::CompileFailed, compiler_result)))
+                    as SFuture<_>
+            } else if cacheable != Cacheable::Yes {
```
This doesn't really *need* to be an else since the if block has a return.

luser commented on src/compiler/compiler.rs:
```diff
@@ -487,15 +474,26 @@ impl Compiler {
     }
 }
 
 /// Write `contents` to `path`.
```
It'd be good to clarify here that this creates a temp dir and just writes a file under *that* path. This is a nice abstraction, though!

luser commented on src/compiler/compiler.rs:
```diff
@@ -504,36 +502,25 @@ pub fn detect_compiler_kind<T>(creator: &T, executable: &str, pool: &CpuPool)
-        Ok((dir, src))
-    });
-    let write = write.chain_err(|| "failed to write temporary file");
+".to_vec();
```
Having to allocate a `Vec` for a static byte string is sort of a bummer, but I doubt it's a big deal with all the other overhead here.

alexcrichton replied:
Ah yeah I assumed this wasn't too bad as well because the compiler info is typically cached anyway (so should hit this rarely)

alexcrichton commented on src/cache/s3.rs:
```diff
@@ -36,39 +35,35 @@ use std::io::{
+        let provider = AutoRefreshingProvider::new(ChainProvider::with_profile_providers(profile_providers, handle));
         //TODO: configurable SSL
-        let bucket = Arc::new(Bucket::new(bucket, endpoint, Ssl::No));
+        let bucket = Rc::new(Bucket::new(bucket, endpoint, Ssl::No, handle));
```
Indeed! Hyper now comes with the [`hyper-tls`](https://github.com/hyperium/hyper-tls) crate which depends on [`native-tls`](https://crates.io/crates/native-tls) which uses the "platform appropriate TLS backend" by default.

alexcrichton commented on src/server.rs:
```diff
@@ -138,15 +139,15 @@ pub struct SccacheServer<C: CommandCreatorSync> {
         // Prepare the service which we'll use to service all incoming TCP
         // connections.
-        let (tx, rx) = mpsc::channel(100);
+        let (tx, rx) = mpsc::channel(1);
```
Oh this was actually an error from before, I intended it to be 1 in the first place and set it to 100 in debugging when I thought something was going wrong, I just forgot to change it back until this commit!

alexcrichton commented:
Ok I think I've addressed everything and cleaned up the commits (but I don't mind cleaning up further if you'd prefer!)

Should be good to go as soon as CI is green.

luser commented on src/commands.rs:
```diff
@@ -101,66 +100,45 @@ fn get_port() -> u16 {
+    });
+
+    let timeout = Duration::from_millis(SERVER_STARTUP_TIMEOUT_MS.into());
+    let timeout = Timeout::new(timeout, &handle)?.map(|()| ServerStartup::TimedOut);
```
This feels like something that could wind up rolled into a convenience method on `Future` in tokio-core, like `fn or_timeout(self, timeout: Duration, handle: &Handle) -> impl Future<Item=TimeoutOr<Self::Item>, Error=Self::Error>`, where `TimeoutOr` could just be like `enum TimeoutOr<T> { Timeout, Value(T) }`.

alexcrichton replied:
Definitely yeah.

On the todo list for the 0.2 release of futures/tokio is to work through a more ergonomic timeout story. For example [`tokio-timer` already has this](https://github.com/tokio-rs/tokio-timer/blob/d803e55b64ab1cb89d43b6802c6fd4494a570115/src/timer.rs#L77-L90) (or similar) and we just need to integrate `tokio-timer` and `tokio_core::reactor::Timeout` more.

alexcrichton commented:
OSX travis seems to be having a lot of issues as of late :(

luser commented:
I was going to try to see if we could get at least one green OS X build but I'm not going to wait on that. I'll land this and build it on my local mac to sanity check. Thanks for all your work here!

alexcrichton commented:
Woo! Thanks for being up for reviewing so much :)

luser commented:
It builds and passes tests on my Mac, so we're fine there.
